index.html

```diff
@@ -787,6 +787,48 @@ <h4>`ThingLink`</h4>
                     </p>
                 </section>
             </section>
+            <section id="exploration-secboot" class="normative">
+                <h3>Security Bootstrapping</h3>
+                <p>The purpose of an exploration service is to serve TDs, but only after suitable
+                authentication, and only to authorized parties.  However, since we don't yet have
+                a TD when accessing an exploration service, we can't depend on the security metadata
+                held in TDs to know what kind of authentication and authorization information is needed.
+                Instead, we have to depend on existing security negotation support in HTTP to
+                bootstrap access, at least to the TD itself.
+                </p>
+                <p>In the HTTP protocol, authentication and authorization is negotiated by
+                the HTTP server returning a "401: Unauthenticated" response code in conjunction with
+                a "WWW-Authentication" header that specifies the information required for
+                authentication.  To gain access, the client then needs to make another request 
+                with the necessary information.
+                <a href="https://www.iana.org/assignments/http-authschemes/http-authschemes.xhtml">There 
+                are several "schemes" for such information registered with IANA.</a>
+                Not all of these are in wide use, some are experimental, and there is only
+                partial overlap with the schemes supported by TDs.
+                To enable security bootstrapping on a wide variety of devices as well
+                as on browsers, the following constraints should be observed:
+                </p>
+                <ul>
+                <li><span class="rfc2119-assertion" id="exploration-secboot-401">Exploration services 
+                MUST reply with an HTTP 401 response code if appropriate
+                authentication information has not been provided.</span>
+                </li>
+                <li><span class="rfc2119-assertion" id="exploration-secboot-auth">A 401 HTTP response 
+                from an exploration service MUST include a <code>WWW-Authentication</code> header
+                describing the required authorizations using one of the following IANA-registered
+                HTTP Authentication Schemes: Basic, Bearer, Digest, or OAuth.
+                </li>
```

Comment:
@farshidtz wrote:
> These kinds of blanket statements aren't appropriate because it may make implementations impossible to comply.

Can you provide an example?

> Especially not when we are so close to publication with little time for testing.

I agree that isn't ideal, but I did highlight this problem a year ago in #135 and I don't really see how the specification can be published without fixing it, because it prevents discovery from working in the common use case where TDs are not public. Publishing public metadata for all the connected devices in a smart home (or in fact any private building) seems like a bad idea...

> I think we are trying to solve a TD limitation by something that isn't really equivalent nor protocol agnostic.

This is a problem that can't really be fixed inside the TD itself, because the problem is with accessing the TD which is part of the discovery process. The 401 solution is certainly HTTP-specific, but for the two exploration mechanisms defined in WoT Discovery 1.0 which both use HTTP, that seems sufficient.

Note that you did previously say:

>> Recommending that for all HTTP interactions is a good idea.
>> This is actually the mandatory approach for OAuth2: https://datatracker.ietf.org/doc/html/rfc6750#section-3

> The data in this header isn't always sufficient for a client to perform authentication.

I highlighted that as an assumption in #135:

>> This also assumes that the WWW-Authenticate header alone provides enough information for a client to authenticate with a directory without the need for any additional out-of-band information (except that provided as part of the authentication process like a user filling out a login form). Hopefully that is the case, but I don't know enough about each of the authentication schemes to be sure. 

If a `WWW-Authentication` header does not provide enough information for a client to be able to authenticate, then we may need to find another solution. However, if that's the case then how does it work in web browsers?

> Other concerns:
> 1. Is the SPARQL engine able to respond 401 with the required `WWW-Authentication`?
> 2. Do SSE clients read or understand `WWW-Authentication` header?
> 3. If the answer to 1 and 2 is no, the assertions should not be for the whole exploration APIs.

The security bootstrapping is only needed for accessing the TD itself. All of these other resources can have security metadata describing them inside the TD.

> 4. What if a server wants to hide a protected endpoint? From [Mozilla's description of HTTP Authentication](https://developer.mozilla.org/en-US/docs/Web/HTTP/Authentication#the_general_http_authentication_framework): "In all cases, the server may prefer returning a 404 Not Found status code, to hide the existence of the page to a user without adequate privileges or not correctly authenticated."

It seems reasonable to allow a 404 as an alternative to a 403 response.

> Is this a state of practice approach for RESTful APIs? Note that the workflows using this header usually involve a user and user-agent (browser).

I don't know. But if an authentication scheme prompts a user for authentication then a user needs to be prompted for authentication, there's not really any way around that. For authentication schemes which require user intervention, my assumption is that the API will need to be consumed by a UI which facilitates that interaction.

FWIW we've just started working on a WoT Directory client in WebThings Cloud and WebThings App which will both consume multiple WebThings Gateways, so we're going to need to find a solution to this problem. I can't guarantee that will be testable in time for the publication timeline though, so it would be great if other implementations could tackle this too.

Reply:
@benfrancis, I think we are agreeing with each other. But the added assersions are not for the TD alone.

> 
> Can you provide an example?

SPAQRL and SSE, as mentioned.

> Note that you did previously say:
> 
> > > Recommending that for all HTTP interactions is a good idea.
> > > This is actually the mandatory approach for OAuth2: https://datatracker.ietf.org/doc/html/rfc6750#section-3
> 

Yes, I said recommending, not mandating.
We can re-mandate it for OAuth2, which would be a redundant assertion.
 
> > Other concerns:
> > 
> > 1. Is the SPARQL engine able to respond 401 with the required `WWW-Authentication`?
> > 2. Do SSE clients read or understand `WWW-Authentication` header?
> > 3. If the answer to 1 and 2 is no, the assertions should not be for the whole exploration APIs.
> 
> The security bootstrapping is only needed for accessing the TD itself. All of these other resources can have security metadata describing them inside the TD.

Well, this PR is mandating it for all HTTP endpoints. To make this only for the bootstrapping of the TD, it has to be part of self-describing only. Note that a directory that exposes its own TD is a self-decribing thing:
> A directory may contain [TDs](https://w3c.github.io/wot-discovery/#dfn-td) and at the same time provide a [TD](https://w3c.github.io/wot-discovery/#dfn-td) and act as a self-describing Thing.
> https://w3c.github.io/wot-discovery/#exploration-overview

Reply:
My intention with these assertions was to ONLY require them for the initial fetch of a TD from an exploration endpoint, not for everything else.  In these cases exploration services just act like "regular" web services except I also limit servers to only use a small subset of well-known and well-supported authentication mechanisms.  This means you can do the whole security-bootstrapping using ordinary web server (and client) libraries.

Good suggestion about allowing a 404 response.  I will also think about the wording of the assertions to make sure they are limited in scope to just the cases where we don't have a TD, e.g. for bootstrapping.  Once you have a TD you can use that instead (although the HTTP transactions still need to be consistent with it).